---
title: "Vignette for `recountmethylation`"
author: "SKM"
date: "12/26/2019"
output:
  pdf_document: default
  html_document: default
---

```{r, eval = T, echo = F}
library(recountmethylation)
```

This vignette describes how to access data using `recountmethylation` and access to the HDF5 database (default file name "remethdb.h5"). This database contains DNA methylation (DNAm) Illumina HM450k array data and sample metadata from the Gene Experession Omnibus.

Version info, file metadata, and other information is accessible from dataset attributes as follows.

```{r}
```

# HDF5 database summary

For convenience and expedient access, the `recountmethylation` library includes functions to access, query, and coerce DNAm array data and sample metadata. DNAm array data is provided in 3 principle types derived from the raw array IDAT files, including the Red and Green raw/unprocessed signal tables, and the noob-normalized Beta-value data table.

# Accessing the HDF5 database

To connect to the databse, establish a path to the principal database file called `remethdb.h5`. You can do this in either of 2 ways, either with local or remote access.

For local access, download the entire HDF5 database file. Note, this file is large (about 150Gb) and download may take awhile.

For remote access, use a connection to the AWS S3 server housing the database file, as follows.

# Querying the HDF5 database

First, access the postprocessed sample metadata with the following.

```{r, eval = T, echo = T}
mdpost = data.mdpost()
```

This returns the postprocessed metadata as a `data.frame` object. You can access the column annotations summaries as follows.

```{r, eval = T, echo = T}
```

Next, apply some filter to narrow the scope of the query. For queries of entire DNAm array data, we recommend limiting the number of unique samples or GSM IDs to <50. You can always implement a query loop to extract and append more than 50 samples across several queries to the database.

Once you have identified samples for your query, you can obtain the DNAm array data in one of two forms, either an `RGChannelSet` object or a `GenomicRatioSet` object. 

The `RGChannelSet` object provides a completely raw/unpreprocessed form of the red and green channel signal data extracted directly from the sample IDAT array image files. From this data, you can readily implement a full preprocessing pipeline, evaluate probe detection p-values, extract methylated or unmethylated signal, etc. Please note the signal data is identified by addresses and not CpG probe IDs at this stage.

The `GenomicRatioSet` returns the noob-normalized Beta-value DNAm data. This dataset corresponds to the Beta-value, or fraction of methylated signal, for each of the 485512 CpG probes on the HM450k BeadArray platform. The noob normalization method is a within-sample out of band signal noramlization that can help remove biases from batch or sample-specific preparations. Finally, we note this data has been rounded for expediency and compactness to 3 decimals. More precise noob-normalized Beta-values are obtainable by starting from queries to the red and green signal tables, described above.

# Session info

```{r, eval = T, echo = T}
sessionInfo()
```